### ISSUE TITLE
xpra crashed "on X error", upgrade killed display

### ISSUE THREAD

**Issue by stdedos** [Xpra-org/xpra#3274]
Before the events that set on action https://github.com/Xpra-org/xpra/issues/3267#issuecomment-922777828, xpra had died unexpectedly.

On the original instance, `display-\:2-20210727-130134.log`:
```
   1978 2021-09-20 12:24:29,650 Warning: client does not support any csc modes with vp9 on window 5417
   1979 /usr/lib/python3/dist-packages/xpra/x11/server_keyboard_config.py:94: Warning: g_object_get_qdata: assertion 'G_IS_OBJECT (object)' failed
   1980   keysym, index = a, b
   1981 /usr/lib/python3/dist-packages/xpra/x11/server_keyboard_config.py:94: Warning: g_object_set_qdata_full: assertion 'G_IS_OBJECT (object)' failed
   1982   keysym, index = a, b
   1983 /usr/lib/python3/dist-packages/xpra/server/window/window_stats.py:177: Warning: g_object_get_qdata: assertion 'G_IS_OBJECT (object)' failed
   1984   encodings_used = [x[1] for x in estats]
   1985 
   1986 (evolution-alarm-notify:65): Gdk-WARNING **: 12:26:32.156: The program 'evolution-alarm-notify' received an X Window System error.
   1987 This probably reflects a bug in the program.
   1988 The error was 'BadDrawable (invalid Pixmap or Window parameter)'.
   1989   (Details: serial 3423 error_code 9 request_code 53 (core protocol) minor_code 0)
   1990   (Note to programmers: normally, X errors are reported asynchronously;
   1991    that is, you will receive the error a while after causing it.
   1992    To debug your program, run it with the GDK_SYNCHRONIZE environment
   1993    variable to change this behavior. You can then get a meaningful
   1994    backtrace from your debugger if you break on the gdk_x_error() function.)
   1995 
   1996 (evolution-calendar-factory:44): evolution-ews-WARNING **: 12:26:33.009: src/EWS/common/e-ews-notification.c:634: Unable to retrieve events for this subscription.  The subscription must be recreated., The events couldn't be read.
   1997 
   1998 
   1999 (evolution.bin:52): mail-notification-CRITICAL **: 12:41:12.167: connection_closed_cb: assertion 'connection != pconnection' failed
   2000 X IO Error
```

then, I thought that "since the xpra died, I might as well upgrade" `display-\:2-20210920-123303.log`:
```
    284 2021-09-20 12:41:09,425 Disconnecting client /run/user/1000/xpra/host-2:
    285 2021-09-20 12:41:09,425  server upgrade
    286 2021-09-20 12:41:09,426 xpra client 1 disconnected.
    287 2021-09-20 12:41:09,426 xpra client 2 disconnected.
    288 2021-09-20 12:41:09,427 xpra client 3 disconnected.
    289 2021-09-20 12:41:09,452 started command 'bash -c xpra id "${DISPLAY}" | grep session-type= | cut -d = -f 2 | grep -qP "shadow" || exit 0 ; env DBUS_SESSION_BUS_ADDRESS="$(cut -f 2- -d= "${XDG_RUNTIME_DIR}/dbus-session")" gnome-screensaver-command -l' with pid 1984411
    290 2021-09-20 12:41:09,468 started command 'bash -c xpra id "${DISPLAY}" | grep session-type= | cut -d = -f 2 | grep -qP "shadow" || exit 0 ; env DBUS_SESSION_BUS_ADDRESS="$(cut -f 2- -d= "${XDG_RUNTIME_DIR}/dbus-session")" xdg-screensaver lock' with pid 1984416
    291 2021-09-20 12:41:11,757 xpra X11 seamless server is exiting
    292 2021-09-20 12:41:11,912 Lost WM selection ManagerSelection(WM_S0), exiting
    293 /usr/lib/python3/dist-packages/gi/overrides/Gtk.py:1632: Warning: g_object_ref: assertion 'G_IS_OBJECT (object)' failed
    294   return _Gtk_main(*args, **kwargs)
    295 2021-09-20 12:41:12,163 killing xvfb with pid 450345
    296 Gdk-Message: 12:41:12.202: Xpra: Fatal IO error 0 (Success) on X server :2.
    297 
    298 Connection failed: [Errno 104] Connection reset by peer
    299 Connection failed: [Errno 104] Connection reset by peer
```

... and subsequently:
```
usr-linux@host [2021-09-20 13:42:38]:/run/user/1000/xpra/2$ cat display-\:2-20210920-124109.log 
2021-09-20 12:41:09,860 cannot access python uinput module:
2021-09-20 12:41:09,860  No module named 'uinput'
2021-09-20 12:41:10,036 pointer device emulation using XTest
2021-09-20 12:41:10,124 serving html content from '/usr/share/xpra/www'
2021-09-20 12:41:10,124 wrote pid 1984427 to '/run/user/1000/xpra/2/server.pid'
2021-09-20 12:41:10,142 found previous Xpra instance
2021-09-20 12:41:10,143 created unix domain socket '/run/user/1000/xpra/host-2'
2021-09-20 12:41:10,144 created unix domain socket '/run/xpra/host-2'
2021-09-20 12:41:10,144 created unix domain socket '/home/usr-linux/.xpra/host-2'
2021-09-20 12:41:10,144 xvfb pid=450345
usr-linux@host [2021-09-20 13:42:51]:/run/user/1000/xpra/2$ tail -v -n+1 display-\:2-20210920-125*
==> display-:2-20210920-125151.log <==
2021-09-20 12:51:51,097 cannot access python uinput module:
2021-09-20 12:51:51,097  No module named 'uinput'
2021-09-20 12:51:54,198 Error: failed to connect to display :2
2021-09-20 12:51:54,198  could not connect to X server on display ':2' after 3 seconds

==> display-:2-20210920-125827.log <==
2021-09-20 12:58:27,218 cannot access python uinput module:
2021-09-20 12:58:27,218  No module named 'uinput'
2021-09-20 12:58:30,218 Error: failed to connect to display :2
2021-09-20 12:58:30,218  could not connect to X server on display ':2' after 3 seconds
usr-linux@host [2021-09-20 13:43:05]:/run/user/1000/xpra/2$
```

**Comment by totaam** [comment]
> `Warning: client does not support any csc modes with vp9 on window 5417`

IIRC, you're running with `--csc-modules=libyuv` - I can't remember why but you will get some warnings like this one.
This should be harmless, it just means that you won't be using `vp9`, no big loss there.

> `Warning: g_object`...

Those are actually quite worrying because there is nothing gobject related in any of those data structures.
So either the warnings are bogus or there is some kind of memory corruption going on.

> `The program 'evolution-alarm-notify' received an X Window System error`

Although this may have been caused by xpra. There isn't much we can do about it.
Evolution should handle this X11 error without crashing.

> then, I thought that "since the xpra died, I might as well upgrade

I see no sign of an xpra server dying in that log... only evolution.

> `Disconnecting client ...`
> ` server upgrade`
> (..)
> `killing xvfb with pid 450345`

This looks like a bug I fixed recently: 9de012dfce5547523d80797779181d2bd5fc43ca
If the server was older than this version... this would be why. Sorry.

**Comment by stdedos** [comment]
> > `Warning: client does not support any csc modes with vp9 on window 5417`
> 
> IIRC, you're running with `--csc-modules=libyuv` - I can't remember why but you will get some warnings like this one.
> This should be harmless, it just means that you won't be using `vp9`, no big loss there.

Issue is https://github.com/Xpra-org/xpra/issues/3167#issuecomment-879752631, and it's kinda open.
The current command line flags seem to help:
```
"Xpra-Python3-x86_64_4.3-r30012\xpra_cmd" attach ssh://usr-linux@1.1.1.1/2 --ssh="plink -ssh -agent" -d ssh --title="@title@ on @@/@server-display@" --modal-windows=no --headerbar=off --use-display=yes --input-method=xim --csc-modules=libyuv --opengl=no
```

> > `Warning: g_object`...
> 
> Those are actually quite worrying because there is nothing gobject related in any of those data structures.
> So either the warnings are bogus or there is some kind of memory corruption going on.

It's either that, _or_ there might be a possibility that there's some bad interaction with the updates.
The server was old, and it must have >6mo uptime 😀 

> > `The program 'evolution-alarm-notify' received an X Window System error`
> 
> Although this may have been caused by xpra. There isn't much we can do about it.
> Evolution should handle this X11 error without crashing.

Ouch 😕
Well, as long as this wasn't xpra crashing and pulling xvfb down with it ...

> > then, I thought that "since the xpra died, I might as well upgrade
> 
> I see no sign of an xpra server dying in that log... only evolution.
> 
> > `Disconnecting client ...`
> > ` server upgrade`
> > (..)
> > `killing xvfb with pid 450345`
> 
> This looks like a bug I fixed recently: [9de012d](https://github.com/Xpra-org/xpra/commit/9de012dfce5547523d80797779181d2bd5fc43ca)
> If the server was older than this version... this would be why. Sorry.

No worries - the server was definitely older than 8 days.
It was due to die at sometime I guess 😕
